### PR TITLE
Replace hand-written lexer with a Gocc generated one

### PR DESCRIPTION
**Pull request by mewmew**

### CODE REVIEW

**Comment by mewmew** [comment]
@awalterschulze Feel free to squash the commit if you deicde to merge the PR. I split it up into several commits to make it easier to review the changes.

Good night :)
/u

**Comment by mewmew** [comment]
Oh, btw. This will probably be a backwards-incompatible PR, since some parts of the API have been redefined or removed altogether (i.e. the `scanner` package is removed entirely).

File: analysewrite_test.go [hunk 1]
```diff
@@ -140,7 +141,7 @@ func TestUnderscore(t *testing.T) {
 }
 
 func TestNonAscii(t *testing.T) {
-	anal(t, `digraph {	label=T�th }`)
+	anal(t, `digraph {	label=Tóth }`)
```
Comment (awalterschulze) [inline analysewrite_test.go]: Did this cause a problem?

Comment (mewmew) [inline analysewrite_test.go]: > Did this cause a problem?

It did indeed. The Gocc generated lexer will return an error stating that the rune is a [Unicode replacement](https://golang.org/pkg/unicode/#ReplacementChar) rune, which signals an error in decoding. Replacing the character with an actual Unicode character resolved the issue.

**Comment by awalterschulze** [comment]
Ok it all looks good except for the merge branch commit.
It has readded files that you have removed in a previous commit.
I don't think these files should exist.

Also you can give yourself some copyright by adding your details to the AUTHORS file.

We can think of the parser and scanner as internal packages which aren't used by the public.  The Parse and Read functions in the top package have not changed and thus I don't think we have broken backwards compatibility.

Thanks again for this HUGE fix, this is some great work :)

**Comment by mewmew** [comment]
> Ok it all looks good except for the merge branch commit.
> It has readded files that you have removed in a previous commit.
> I don't think these files should exist.

Ok, I'll look into reverting the merge commit and making a new merge against master where these files are not included. I'm at my cousin's wedding on the west coast of Sweden, so give me a couple of days :)

> We can think of the parser and scanner as internal packages which aren't used by the public. The Parse and Read functions in the top package have not changed and thus I don't think we have broken backwards compatibility.

I would reckon so too. Perhaps we may look into moving the generated code into an `internal` directory, so that it is made clear that this API is not intended for users? This is the approach I took with the dot package. https://github.com/graphism/dot/tree/master/internal

> Thanks again for this HUGE fix, this is some great work :)

You are more than welcome. Felt happy to contribute something back to this project, as it has helped me a lot in the past and been a huge source of insipration.

**Comment by mewmew** [comment]
> Also you can give yourself some copyright by adding your details to the AUTHORS file.

The grammar for the lexer was copied verbatim from the dot project, which is already released into the public domain. So, I claim no copyright for the added code. I hope this is ok?

**Comment by awalterschulze** [comment]
I am just saying there is an opportunity to claim some copyright, but you can always just add yourself to the CONTRIBUTORS file.  This just says you have contributed, but don't claim any copyright.  No problem either is fine :)

Take your time, enjoy the wedding :)

I am glad to hear something I did could be an inspiration to you.

**Comment by awalterschulze** [comment]
The Authors file includes the diff that is already in master.
```
+Xavier Chassin <xavier.chassin@live.fr>
```
And there is more left.

Glad you had a great wedding :)
No worries on the delay.

**Comment by mewmew** [comment]
> The Authors file includes the diff that is already in master.

Sorry, still figuring out how to use git merge :)

> And there is more left.

Could you elaborate?

**Comment by awalterschulze** [comment]
Sorry what I mean with more left is just more files that were not part of your original diff.
This is only a git merge issue, but the rest of it is perfect.

**Comment by awalterschulze** [comment]
Thanks again for this.
After investigating it seems that only the github view of this pull request was weird.
The git diff between the branches were completely sane.
So I decided to merge it :)

**Comment by mewmew** [comment]
Lovely to hear :) Let me know if any issues arries related to the Gocc generated lexer.

**Comment by awalterschulze** [comment]
Will do thank you very much for all this work.  I think this is much better than the solution I had :)

**Pull request merged**